vm/interpreter/src/circ_supply.rs

```diff
@@ -0,0 +1,254 @@
+    let reward_actor = get_actor_state(state_tree, &*REWARD_ACTOR_ADDR)?;
+    let reward_state: reward::State = state_tree
+        .store()
+        .get(&reward_actor.state)?
+        .ok_or_else(|| "Failed to get Rewrad Actor State".to_string())?;
+
+    Ok(reward_state.total_storage_power_reward().clone())
+}
+
+pub fn get_fil_market_locked<DB: BlockStore>(
+    state_tree: &StateTree<DB>,
+) -> Result<TokenAmount, Box<dyn StdError>> {
+    let market_actor = get_actor_state(state_tree, &*STORAGE_MARKET_ACTOR_ADDR)?;
+
+    let market_state: market::State = state_tree
+        .store()
+        .get(&market_actor.state)?
+        .ok_or_else(|| "Failed to get Market Actor State".to_string())?;
+
+    Ok(market_state.total_locked())
+}
+
+pub fn get_fil_power_locked<DB: BlockStore>(
+    state_tree: &StateTree<DB>,
+) -> Result<TokenAmount, Box<dyn StdError>> {
+    let power_actor = get_actor_state(state_tree, &*STORAGE_POWER_ACTOR_ADDR)?;
+
+    let power_state: power::State = state_tree
+        .store()
+        .get(&power_actor.state)?
+        .ok_or_else(|| "Failed to get Power Actor State".to_string())?;
+
+    Ok(power_state.total_locked().clone())
+}
+
+pub fn get_fil_reserve_disbursed<DB: BlockStore>(
+    state_tree: &StateTree<DB>,
+) -> Result<TokenAmount, Box<dyn StdError>> {
+    let power_actor = get_actor_state(state_tree, &RESERVE_ADDRESS)?;
+
+    // If money enters the reserve actor, this could lead to a negative term
+    Ok(FIL_RESERVED.clone() - power_actor.balance)
+}
+
+pub fn get_fil_locked<DB: BlockStore>(
+    state_tree: &StateTree<DB>,
+) -> Result<TokenAmount, Box<dyn StdError>> {
+    let market_locked = get_fil_market_locked(&state_tree)?;
+    let power_locked = get_fil_power_locked(&state_tree)?;
+    Ok(power_locked + market_locked)
+}
+
+pub fn get_fil_burnt<DB: BlockStore>(
+    state_tree: &StateTree<DB>,
+) -> Result<TokenAmount, Box<dyn StdError>> {
+    let burnt_actor = get_actor_state(state_tree, &*BURNT_FUNDS_ACTOR_ADDR)?;
+
+    Ok(burnt_actor.balance)
+}
+
+pub fn get_circulating_supply<'a, DB: BlockStore>(
+    pre_ignition: &GenesisInfo,
+    post_ignition: &GenesisInfo,
+    height: ChainEpoch,
+    state_tree: &StateTree<'a, DB>,
+) -> Result<TokenAmount, Box<dyn StdError>> {
+    let fil_vested = get_fil_vested(&pre_ignition, &post_ignition, height, &state_tree)?;
+    let fil_mined = get_fil_mined(&state_tree)?;
+    let fil_burnt = get_fil_burnt(&state_tree)?;
+    let fil_locked = get_fil_locked(&state_tree)?;
+    let fil_reserve_distributed = if height > UPGRADE_ACTORS_V2_HEIGHT {
+        get_fil_reserve_disbursed(&state_tree)?
+    } else {
+        TokenAmount::default()
+    };
+    let fil_circulating = BigInt::max(
+        &fil_vested + &fil_mined + fil_reserve_distributed - &fil_burnt - &fil_locked,
+        TokenAmount::default(),
+    );
+
+    Ok(fil_circulating)
+}
+
+fn get_totals_by_epoch() -> HashMap<ChainEpoch, TokenAmount> {
```

Comment:
If you actually need to use a map (I assume this is not necessary), then the return should be an AHashMap or FnvHashMap, it's an integer key which these hash functions are much faster for

Checking usage it seems like this can just be a tuple vector?

Reply:
yeah, looks like this can be a static ref vec

Reply:
making it a static vec